cylc/uiserver/authorise.py

```diff
@@ -0,0 +1,477 @@
+        # Now we have a reduced site auth dictionary for the current owner
+        return owner_dict
+
+    def return_site_auth_defaults_for_access_user(
+            self, access_user: Dict[str, Union[str, Sequence[Any]]]) -> Set:
+        """Return site authorization defaults for given access user.
+            Args:
+                access_user: access_user dictionary, in the form
+                            {'access_username': username
+                             'access_user_group: [group1, group2,...]'
+                            }
+            Returns:
+                Set of default operations permitted
+
+        """
+        defaults: Set[str] = set()
+        if not self.owner_dict:
+            return defaults
+        items_to_check = ['*', access_user['access_username']]
+        items_to_check.extend(access_user['access_user_groups'])
+        for item in items_to_check:
+            permission = ''
+            with suppress(KeyError):
+                permission = self.owner_dict[item].get(
+                    Authorization.DEFAULT, ''
+                )
+            if isinstance(permission, str):
+                defaults.add(permission)
+            else:
+                defaults.update(permission)
+        defaults.discard('')
+        return defaults
+
+
+# GraphQL middleware
+class AuthorizationMiddleware:
+
+    auth = None
+    current_user = None
+
+    def resolve(self, next_, root, info, **args):
+        # We won't be re-checking auth for return variables
+        if len(info.path) > 1:
+            return next_(root, info, **args)
+        op_name = self.get_op_name(info.field_name, info.operation.operation)
+        # It shouldn't get here but worth checking for zero trust
+        if not op_name:
+            self.auth_failed(self.current_user, op_name, 400,
+                             "Operation not in schema.")
+        authorised = self.auth.is_permitted(self.current_user, op_name)
+        if not authorised:
+            self.auth_failed(self.current_user, op_name, 403)
+        if (info.operation.operation in Authorization.ASYNC_OPS
+                or iscoroutinefunction(next_)):
+            return self.async_resolve(next_, root, info, **args)
+        return next_(root, info, **args)
+
+    @staticmethod
+    def auth_failed(current_user: str, op_name: str,
+                    http_code: int, message: Optional[str] = None):
+        """
+        Raise authorization error
+        Args:
+            current_user: username accessing operation
+            op_name: operation name
+            http_code: http error code to raise
+            message: Message to log Defaults to None.
+
+        Raises:
+            web.HTTPError
+        """
+        log_message = (f"Authorization failed for {current_user}"
+                       f":requested to {op_name}.")
+        if message:
+            log_message = log_message + " " + message
+        logger.warn(log_message)
```

Comment:
Just wondering if we should log authorization errors a bit different... if I am responsible for a webapp, I like to find a quick and simple way to find all the audit logs, where I can see things like authorization errors (similarly to crash logs, or trace logs, etc...).

Maybe always include some piece of text in each authorization error message, so users can easily grep for it (and we can maybe filter in the future Cylc Review replacement?).

Or maybe include an error number, or log to a different logger?

Reply:
Yes, good idea. I'll have a think and do some reading and see if I can come up with an alternative.

Reply:
@datamel looks like `logger.warn` is deprecated (IDE warning about the warn :smile: :drum: ). I think instead we must use `logger.warning`.